File: catalogue-application/src/test/java/io/quartic/catalogue/datastore/GoogleDatastoreBackendShould.java
package io.quartic.catalogue.datastore;

import com.google.cloud.datastore.Datastore;
import com.google.cloud.datastore.Entity;
import com.google.cloud.datastore.Key;
import com.google.cloud.datastore.PathElement;
import com.google.cloud.datastore.testing.LocalDatastoreHelper;
import io.quartic.catalogue.StorageBackend;
import io.quartic.catalogue.StorageBackendTests;
import io.quartic.catalogue.api.model.DatasetConfig;
import io.quartic.catalogue.api.model.DatasetCoordinates;
import org.joda.time.Duration;
import org.junit.AfterClass;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;

import java.io.IOException;
import java.util.Map;
import java.util.concurrent.TimeoutException;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.empty;
import static org.hamcrest.Matchers.equalTo;

public class GoogleDatastoreBackendShould extends StorageBackendTests {
    private static LocalDatastoreHelper helper;
    private Datastore datastore;
    private GoogleDatastoreBackend backend;

    @BeforeClass
    public static void setupEmulator() throws IOException, InterruptedException {
        helper = LocalDatastoreHelper.create();
        helper.start();
    }

    @AfterClass
    public static void tearDownEmulator() throws InterruptedException, TimeoutException, IOException {
        helper.stop(Duration.millis(3000));
    }

    @Before
    public void setUp() throws IOException, InterruptedException {
        helper.reset();
        datastore = helper.getOptions()
                .toBuilder()
                .setNamespace("test")
                .build().getService();
        backend = new GoogleDatastoreBackend(datastore);
    }

    @Override
    protected StorageBackend getBackend() {
        return backend;
    }

    @Test
    public void respect_datastore_namespace_separation() throws IOException {
        GoogleDatastoreBackend secondBackend = new GoogleDatastoreBackend(helper.getOptions()
                .toBuilder()
                .setNamespace("test2")
                .build().getService());

        backend.put(coords("namespace", "A"), dataset("1"));
        secondBackend.put(coords("namespace", "A"), dataset("2"));

        Map<DatasetCoordinates, DatasetConfig> datasets = backend.getAll();
        Map<DatasetCoordinates, DatasetConfig> secondDatasets = secondBackend.getAll();

        assertThat(datasets.size(), equalTo(1));
        assertThat(secondDatasets.size(), equalTo(1));

        assertThat(datasets.get(coords("namespace", "A")).getMetadata().getName(), equalTo("1"));
        assertThat(secondDatasets.get(coords("namespace", "A")).getMetadata().getName(), equalTo("2"));
    }

    @Test
    public void ignore_unnamespaced_datasets() throws Exception {
        // This is where a v1 dataset would be
        final Key key = datastore.newKeyFactory()
                .addAncestors(PathElement.of("catalogue", "ancestor"))
                .setKind("dataset")
                .newKey("foobles");
        datastore.put(Entity.newBuilder(key).build());

        assertThat(backend.getAll().entrySet(), empty());
    }
}

File: catalogue/src/test/kotlin/io/quartic/catalogue/datastore/GoogleDatastoreBackendShould.kt
package io.quartic.catalogue.datastore

import com.google.cloud.datastore.Datastore
import com.google.cloud.datastore.Entity
import com.google.cloud.datastore.PathElement
import com.google.cloud.datastore.testing.LocalDatastoreHelper
import io.quartic.catalogue.StorageBackendTests
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.empty
import org.hamcrest.Matchers.equalTo
import org.joda.time.Duration
import org.junit.AfterClass
import org.junit.Before
import org.junit.BeforeClass
import org.junit.Test

class GoogleDatastoreBackendShould : StorageBackendTests() {
    private lateinit var datastore: Datastore
    lateinit override var backend: GoogleDatastoreBackend

    @Before
    fun setUp() {
        helper!!.reset()
        datastore = helper!!.options
                .toBuilder()
                .setNamespace("test")
                .build().service
        backend = GoogleDatastoreBackend(datastore)
    }

    @Test
    fun respect_datastore_namespace_separation() {
        val secondBackend = GoogleDatastoreBackend(helper!!.options
                .toBuilder()
                .setNamespace("test2")
                .build().service)

        backend[coords("namespace", "A")] = dataset("1")
        secondBackend[coords("namespace", "A")] = dataset("2")

        val datasets = backend.getAll()
        val secondDatasets = secondBackend.getAll()

        assertThat(datasets.size, equalTo(1))
        assertThat(secondDatasets.size, equalTo(1))

        assertThat(datasets[coords("namespace", "A")]?.metadata?.name, equalTo("1"))
        assertThat(secondDatasets[coords("namespace", "A")]?.metadata?.name, equalTo("2"))
    }

    @Test
    fun ignore_unnamespaced_datasets() {
        // This is where a v1 dataset would be
        val key = datastore.newKeyFactory()
                .addAncestors(PathElement.of("catalogue", "ancestor"))
                .setKind("dataset")
                .newKey("foobles")
        datastore.put(Entity.newBuilder(key).build())

        assertThat(backend.getAll().entries, empty())
    }

    companion object {
        private var helper: LocalDatastoreHelper? = null

        @BeforeClass
        @JvmStatic
        fun setupEmulator() {
            helper = LocalDatastoreHelper.create()
            helper!!.start()
        }

        @AfterClass
        @JvmStatic
        fun tearDownEmulator() {
            helper!!.stop(Duration.millis(3000))
        }
    }
}
